lib/Dialect/RTG/IR/RTGTypes.cpp

```diff
@@ -19,6 +19,77 @@ using namespace rtg;
 #define GET_TYPEDEF_CLASSES
 #include "circt/Dialect/RTG/IR/RTGTypes.cpp.inc"
 
+//===----------------------------------------------------------------------===//
+// DictType
+//===----------------------------------------------------------------------===/
+
+LogicalResult DictType::verify(function_ref<InFlightDiagnostic()> emitError,
+                               ArrayRef<StringAttr> entryNames,
+                               ArrayRef<Type> entryTypes) {
+  if (entryNames.size() != entryTypes.size())
+    return emitError() << "must have the same number of names and types";
+
+  llvm::SmallDenseSet<StringAttr> entryNameSet;
+  LogicalResult result = success();
+  entryNameSet.reserve(entryNames.size());
+  for (auto name : entryNames) {
+    if (!entryNameSet.insert(name).second) {
+      result = failure();
+      emitError() << "duplicate entry name '" << name.getValue()
+                  << "' in rtg.dict type";
+    }
+  }
+  return result;
+}
+
+Type DictType::parse(AsmParser &p) {
+  llvm::StringSet<> nameSet;
+  SmallVector<Type> types;
+  SmallVector<StringAttr> names;
+  bool hasDuplicateName = false;
+
+  auto parseResult = p.parseCommaSeparatedList(
+      mlir::AsmParser::Delimiter::LessGreater, [&]() -> ParseResult {
+        std::string name;
+        Type type;
+
+        auto fieldLoc = p.getCurrentLocation();
+        if (p.parseKeywordOrString(&name) || p.parseColon() ||
+            p.parseType(type))
+          return failure();
+
+        if (!nameSet.insert(name).second) {
+          p.emitError(fieldLoc, "duplicate entry name \'" + name + "\'");
+          // Continue parsing to print all duplicates, but make sure to error
+          // eventually
+          hasDuplicateName = true;
+        }
+
+        types.push_back(type);
+        names.push_back(StringAttr::get(p.getContext(), name));
+        return success();
+      });
+
+  if (hasDuplicateName)
+    return Type();
+
+  if (failed(parseResult))
+    return Type();
+
+  return get(p.getContext(), names, types);
```

Comment:
Do we need to sort the fields by name, so that we can guarantee that two dictionary types with the same fields intern to the same thing?